design/pipeline-caching.md

```diff
@@ -0,0 +1,157 @@
+
+The caching strategy would default to  dedup file transfer mechanism but allow an option to override the caching strategy with another approach, for example"
+
+Selecting the cache strategy would simply be a matter of specifying a strategy on the task inputs (for both the restore and save tasks, restore shown below):
+
+```yaml
+steps:
+- task: RestoreCache@0
+  inputs:
+    strategy: dedup
+    key: |
+      package.json
+      yarn.lock
+    paths: |
+      node_modules
+```
+
+### YAML syntax
+
+The example task usage above uses the explicit task references. We also want to provide a streamlined syntax for caching, applied to the example above.
+
+```yaml
+steps:
+- restoreCache: yarn.lock
+- script: yarn
+- saveCache:
+  key: yarn.lock
+  paths: node_modules
+- script: yarn build
+```
+
+This more compact syntax just reduces some of the noise around using the caching task. Additionally we are planning to integrate caching into the the ```use:``` syntax. This will look like the following:
+
+```yaml
+steps:
+- use: node
+  cache: true
+- script: |
+    yarn
+    yarn build
+    yarn test
+```
+
+The idea behind the ```use:``` syntax is that given what we are using (in this case ```node```) we make some automatic decisions about what and how to cache. This logic will use various heuristics to make the decision about what to cache, for example, it will look for the presence of a ```yarn.lock``` file or a ```package-lock.json``` to determine where the ```node_modules``` path may be located.
+
+Other ```use:``` statements will have similar heuristics to pick the best defaults.
+
+Note that the ```use:``` syntax will inject the cache save step at the end of the build process which will not always be desirable. In those cases developers can fall back to specifying the ```restorCache:``` and ```saveCache:``` YAML statements.
+
+### Cache Scoping
+
+Getting scoping right is important for maximising cache hits and but also avoiding the cache becoming an attack vector to insert malicous code into official/master builds.
+
+As a result we will automatically scope caches to the branch that they are running against. The caches will also be hierarchical so a feature branch will be able to get a hit on the master, but when populatingn the cache in the ```saveCache:``` step, the contents of that cache won't be used on the master build.
+
+For PR builds, the PR build will use the cache of the branch it is merging into or from with preference been given to the branch it is merging from. At this stage, PR builds, regardless of whether they are from a branch within the repo or from a forked clone will not be able to store content in the cache.
+
+### Cache Expiry
+
+Cache lifetime will be best effort. Our underlying storage will generally keep content for 7 days before it becomes a candidate for eviction, but we won't initially make a guarantee here. We will evaluate the effectiveness of cache durations and listen to community feedback.
```

Comment:
Control mechanisms on cache life cycle would be beneficial, such as the ability to set an age limit on the cache, and trigger a new hydration of the cache (perhaps by triggering a build on cache-expiry)

Reply:
For V1 we are going to try and keep it simple and have a default 7 day expiry on the cache duration but be explicitly best effort. We'll provide the ability to force a cache miss via salting (and certain scenarios will automatically trigger a miss).

Could you go into more detail about what you mean by trigger a new hydration of the cache? It seems like you are talking about forcing a miss and new cache creation but I'm not 100% sure.